Atlantis unlock and discard all plans via VCS comment

parmouraly opened the pull request:
**Objective**
In order to allow someone to discard a plans and `Atlantis` locks without having to close/reopen their PR, we would like to add upstream functionality that enables devs to discard plans via VCS comments in an analogous way to planning/applying.

**Context**
The ability to discard an atlantis plan on demand is useful particularly for long lived PRs after a successful plan, as they would tend to block other people from performing Atlantis plan on the same project(s) for as long as the original PR is still open and successfully planned, which is not optimal for parallel working.

Atlantis provides a UI action that allows you to discard a plan and remove the atlantis lock.
However for security reasons because of the access Atlantis agents have to modify infrastructure, the UI might not always be reachable due to firewall restrictions.

lkysow commented:
Hi @parmouraly thanks for the PR!
I don't think you need to follow the pattern for the other commands with the command_builder. Instead I think we should take the code from the `DeleteLock` method `locks_controller` and factor it out into a new struct `DeleteLockCommand`. Then both the `DeleteLock` and command_runner can call this command struct.

parmouraly commented:
Hi @lkysow thanks for the quick feedback!
That makes sense, I will work on that and push a follow up between today and tomorrow.

On a more philosophical level one thing I was wondering about is the usage of the keyword `discard` vs smth more light like `unlock`.
I did debate which of the two would be better. `Unlock` feels a bit more to the point and less scary to me but I used `discard` be following the convention in the `LockController`.

parmouraly commented:
Got side-tracked with looking into a bug for which I've put a PR up.
Is the refactor in my last commit moving in the right direction?
Hoping to wrap up the final bits tomorrow.

matthewellis-pp commented:
Would love to see some movement on this as its a very important feature for me, please reach out if i can be of any assistance on this.

parmouraly commented:
hey @matthewellis-pp 👋 
I've been caught up with some work. I think @lkysow is onboard with that one but he's also been busy. 
I'm going to try and wrap this up this week code wise, there is not much left to do. And hopefully we can get this reviewed and merged soon.

lkysow commented:
Hi @parmouraly so you've got closer to what I was asking for.

I don't think we need to change project_command_builder. I think we should be able to call `DeleteLockCommand` directly from `command_runner`:

```diff
	if !c.validateCtxAndComment(ctx) {
		return
	}

+	if cmd.CommandName() == models.DiscardCommand {
+		c.DeleteLockCommand.Delete(cmd)
+		l.VCSClient.CreateComment(lock.Pull.BaseRepo, lock.Pull.Num, comment)
+		return
+	}

	if cmd.CommandName() == models.ApplyCommand {
```

lkysow commented on .gitignore:
```diff
@@ -12,4 +12,5 @@ node_modules/
-atlantis
\ No newline at end of file
+atlantis
+temp.sh
```
why?

parmouraly replied:
oops that was unintentional. Removed

lkysow commented on CONTRIBUTING.md:
```diff
@@ -55,7 +55,7 @@ docker run --rm -v $(pwd):/go/src/github.com/runatlantis/atlantis -w /go/src/git
 ## Calling Your Local Atlantis From GitHub
 - Create a test terraform repository in your GitHub.
-- Create a personal access token for Atlantis. See [Create a GitHub token](https://github.com/runatlantis/atlantis#create-a-github-token).
+- Create a personal access token for Atlantis. See [Create a GitHub token](https://github.com/runatlantis/atlantis/tree/master/runatlantis.io/docs/access-credentials.md#generating-an-access-token).
```
thanks!

parmouraly replied:
No problem!

lkysow commented on server/events/comment_parser.go:
```diff
@@ -61,6 +61,8 @@ type CommentBuilder interface {
 	// BuildApplyComment builds an apply comment for the specified args.
 	BuildApplyComment(repoRelDir string, workspace string, project string) string
+	// BuildDiscardComment builds a discard comment for the specified args.
+	BuildDiscardComment(repoRelDir string, workspace string, project string, commentArgs []string) string
```
Shouldn't need commentArgs

parmouraly replied:
Removed

lkysow commented on server/events/comment_parser.go:
```diff
@@ -157,7 +159,7 @@ func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType) Commen
 	}
 
 	// Need to have a plan or apply at this point.
```
```suggestion
	// Need to have a plan, apply or discard at this point.
```

lkysow commented on server/events/comment_parser.go:
```diff
@@ -186,6 +188,14 @@ func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType) Commen
+		name = models.DiscardCommand
+		flagSet = pflag.NewFlagSet(models.DiscardCommand.String(), pflag.ContinueOnError)
+		flagSet.SetOutput(ioutil.Discard)
+		flagSet.StringVarP(&workspace, workspaceFlagLong, workspaceFlagShort, "", "Switch to this Terraform workspace before planning.")
```
```suggestion
		flagSet.StringVarP(&workspace, workspaceFlagLong, workspaceFlagShort, "", "Discard the plan for this workspace.")
```

lkysow commented on server/events/comment_parser.go:
```diff
@@ -186,6 +188,14 @@ func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType) Commen
+		flagSet = pflag.NewFlagSet(models.DiscardCommand.String(), pflag.ContinueOnError)
+		flagSet.SetOutput(ioutil.Discard)
+		flagSet.StringVarP(&workspace, workspaceFlagLong, workspaceFlagShort, "", "Switch to this Terraform workspace before planning.")
+		flagSet.StringVarP(&dir, dirFlagLong, dirFlagShort, "", "Which directory to run plan in relative to root of repo, ex. 'child/dir'.")
```
```suggestion
		flagSet.StringVarP(&dir, dirFlagLong, dirFlagShort, "", "Discard the plan for this directory.")
```

parmouraly replied:
Done, although if we go with `unlock` for the whole PR to begin with then we don't need this bit.

lkysow commented on server/events/comment_parser.go:
```diff
@@ -264,6 +274,22 @@ func (e *CommentParser) BuildApplyComment(repoRelDir string, workspace string, p
 }
 
+// BuildDiscardComment builds discard comment for the specified args.
+func (e *CommentParser) BuildDiscardComment(repoRelDir string, workspace string, project string, commentArgs []string) string {
```
won't need this

parmouraly replied:
Removed

lkysow commented on server/events/comment_parser.go:
```diff
@@ -342,11 +368,13 @@ Examples:
+           To plan a specific project, use the -d, -w and -p flags.
+  apply    Runs 'terraform apply' on all unapplied plans from this pull request.
+           To only apply a specific plan, use the -d, -w and -p flags.
+  discard  Discards all plans in this PR as well as the atlantis locks.
```
If you support -d, -w and -p then this should read.

```
Discards all plans in this PR and deletes all locks held by this PR.
To discard a specific project, use the -d, -w and -p flags.
```

parmouraly replied:
I was thinking, for simplicity, to only support general discard/unlock for the whole PR in this version. Would that be OK?

lkysow replied:
Yes that sounds good.

lkysow commented on server/events/delete_lock_command.go:
```diff
@@ -0,0 +1,84 @@
+// Copyright 2017 HootSuite Media Inc.
```
Don't need this boilerplate because this is a new file. That's only for old files.

parmouraly replied:
Removed

lkysow commented on server/events/delete_lock_command.go:
```diff
@@ -0,0 +1,84 @@
+			l.Logger.Err("unable to delete project status: %s", err)
+		}
+	}
+}
```
Need tests for this file.

parmouraly replied:
Tests coming

parmouraly replied:
Tests added 👍

lkysow commented on server/events/markdown_renderer.go:
```diff
@@ -24,8 +24,9 @@ import (
-	applyCommandTitle = "Apply"
+	planCommandTitle    = "Plan"
+	applyCommandTitle   = "Apply"
+	discardCommandTitle = "Discard"
```
I don't think you need to modify this renderer. Should be able to construct this comment in command_runner.

parmouraly replied:
Done. I left the bit that shows the option for discarding plans

parmouraly replied:
I mean after a successful plan, you get options to click to unlock, apply etc, so I added the unlock option

parmouraly commented:
Just rebased to master

parmouraly commented on server/events/delete_lock_command_test.go:
```diff
@@ -0,0 +1,26 @@
+	. "github.com/runatlantis/atlantis/testing"
+)
+
+func TestDeleteLock_LockerErr(t *testing.T) {
```
I made a start, more tests are coming

parmouraly replied:
added

parmouraly commented on server/testfixtures/test-repos/simple/exp-output-autoplan.txt.act:
```diff
@@ -0,0 +1,42 @@
+* :arrow_forward: To **apply** this plan, comment:
+    * `atlantis apply -d .`
+* :put_litter_in_its_place: To **delete** this plan click [here](lock-url), or to delete all plans comment and atlantis locks:
+    * `atlantis unlock`
```
How does this look @lkysow . Are you happy with this keyword? I can adapt the comment parser if so

parmouraly commented on server/locks_controller_test.go:
```diff
@@ -140,129 +135,23 @@ func TestDeleteLock_InvalidLockID(t *testing.T) {
 	responseContains(t, w, http.StatusBadRequest, "Invalid lock id \"%A@\"")
 }
 
+// TODO: I'm going to adapt the next 5 test cases to use some mock of DeleteLockCommand
```
Am I OK to generate mocks for `DeleteLockCommand` so I can refactor these tests?

lkysow replied:
Yes, looks like that will be too hard to instantiate a concrete instance of.

parmouraly commented:
@lkysow I think so far I've addressed all your comments.

I think the three main outstanding points are:
* Reach an agreement on providing only a general `atlantis unlock` at this first iteration, for the whole PR. Just to keep things simple and take it step by step. Are we good with `unlock`?
* Some test additions/adaptations to maintain or improve coverage
* Deciding on whether to generate mocks for the new `DeleteLockCommand` in order to restore the tests in the `locks_controller_test`. Should I go for it, or do you prefer to avoid mocks here too?

Is there anything else I'm forgetting? Do you think we are close to merging this one?
If not what can I do to get this in.

lkysow commented on server/events/comment_parser.go:
```diff
@@ -186,6 +186,14 @@ func (e *CommentParser) Parse(comment string, vcsHost models.VCSHostType) Commen
+		name = models.DiscardCommand
+		flagSet = pflag.NewFlagSet(models.DiscardCommand.String(), pflag.ContinueOnError)
+		flagSet.SetOutput(ioutil.Discard)
+		flagSet.StringVarP(&workspace, workspaceFlagLong, workspaceFlagShort, "", "Discard the plan for this workspace.")
```
you'll need to remove the flags -p, etc. here if you're only doing the whole PR

parmouraly commented:
I've converted this back to draft because I'm missing quite a few tests still... Coming soon

parmouraly commented:
@lkysow this is a working version 🎉 
As in it seems to work fine when manually testing on my machine
I think I've addressed all your comments so far. Please let me know if I can cover anything else.

In the mean time I'll see if I can push a couple more unit tests for `models` and `comment_parser`

lkysow commented on server/events/delete_lock_command.go:
```diff
@@ -0,0 +1,79 @@
+			l.Logger.Err("unable to delete project status: %s", err)
+		}
+	}
+}
```
great work on this command.

lkysow commented on server/events/markdown_renderer.go:
```diff
@@ -249,7 +248,8 @@ var planSuccessWrappedTmpl = template.Must(template.New("").Parse(
 	"    * `{{.ApplyCmd}}`\n" +
-	"* :put_litter_in_its_place: To **delete** this plan click [here]({{.LockURL}})\n" +
+	"* :put_litter_in_its_place: To **delete** this plan click [here]({{.LockURL}}), or to delete all plans and atlantis locks comment:\n" +
+	"    * `atlantis unlock`\n" +
```
I'm going to move this to the bottom where we talk about running atlantis apply to apply all plans because unlocking all plans makes more sense there than nested with each individual plan

parmouraly replied:
Ah yea good point didn't think about it like that. I just kept it in the already existing `delete` section to avoid adding a new bullet point but yea sounds good 👍

parmouraly replied:
It's just gonna be a bit a of a pain to update the `rendere_tests` :-D

lkysow commented on server/events/models/models.go:
```diff
@@ -522,6 +524,20 @@ func (c CommandName) String() string {
 }
+
+// DeleteLockCommandResult is the result of attempting to delete an atlantis lock.
+type DeleteLockCommandResult int
```
I've made some changes that make this unnecessary, mainly following the pattern before where `error` means error and `nil` means lock not found.

lkysow commented:
Merged with #1091! Great work 🎉
